--- a/exfat-utils/libexfat/io.c
+++ b/exfat-utils/libexfat/io.c
@@ -322,7 +322,7 @@ off_t exfat_seek(struct exfat_dev* dev,
 	/* XXX SEEK_CUR will be handled incorrectly */
 	return dev->pos = lseek(dev->fd, offset, whence);
 #else
-	return lseek(dev->fd, offset, whence);
+	return lseek64(dev->fd, offset, whence);
 #endif
 }
 
@@ -356,7 +356,7 @@ ssize_t exfat_pread(struct exfat_dev* de
 #ifdef USE_UBLIO
 	return ublio_pread(dev->ufh, buffer, size, offset);
 #else
-	return pread(dev->fd, buffer, size, offset);
+	return pread64(dev->fd, buffer, size, offset);
 #endif
 }
 
@@ -366,7 +366,7 @@ ssize_t exfat_pwrite(struct exfat_dev* d
 #ifdef USE_UBLIO
 	return ublio_pwrite(dev->ufh, buffer, size, offset);
 #else
-	return pwrite(dev->fd, buffer, size, offset);
+	return pwrite64(dev->fd, buffer, size, offset);
 #endif
 }
 
