diff -Naur samba/Makefile samba-brcm/Makefile
--- samba/Makefile	2010-09-14 20:04:24.000000000 -0300
+++ samba-brcm/Makefile	2018-11-24 14:14:21.272106657 -0400
@@ -1,14 +1,18 @@
 
 all:
 	if [ ! -e ./source/Makefile ];then \
-	cd ./source; ./configure --host=mips-linux-uclibc --target=mips-linux-uclibc --disable-developer --enable-cups=no --enable-fam=no --enable-shared=yes --enable-iprint=no --enable-pie=no --without-ldap --without-cifsupcall --without-cifsmount --without-cluster-support --without-acl-support --with-privatedir=/var/samba --with-lockdir=/var/samba/locks --with-configdir=/var/samba --with-piddir=/var/samba --with-logfilebase=/var/samba SMB_BUILD_CC_NEGATIVE_ENUM_VALUES=no linux_getgrouplist_ok=no; \
+	cd ./source; ./configure --host=mips-linux-uclibc --target=mips-linux-uclibc --disable-developer --enable-cups=no --enable-fam=no --enable-shared=yes --enable-iprint=no --enable-pie=no --without-ldap --without-cifsupcall --without-cifsmount --without-cluster-support --without-acl-support --with-privatedir=/var/samba --with-lockdir=/var/samba/locks --with-configdir=/var/samba --with-piddir=/var/samba --with-logfilebase=/var/samba SMB_BUILD_CC_NEGATIVE_ENUM_VALUES=no linux_getgrouplist_ok=no samba_cv_have_longlong=yes samba_cv_HAVE_IFACE_IFCONF=yes ; \
 	fi
 
-	$(MAKE) -C ./source;
+	$(MAKE) -C ./source srcdir=./ builddir=./
 	install -m 755 ./source/bin/smbd $(INSTALL_DIR)/bin/smbd
 	install -m 755 ./source/bin/smbpasswd $(INSTALL_DIR)/bin/smbpasswd
-	cp ./source/conf/smb.conf $(INSTALL_DIR)/etc
+ifneq ($(strip $(SUPPORT_GPL_UNDEFINED)),)
+	install -m 444 ./source/conf/smb_telus.conf $(INSTALL_DIR)/etc/smb.conf
+else
+	install -m 444 ./source/conf/smb.conf $(INSTALL_DIR)/etc
+endif
 
 clean:
-	if [ -f ./source/Makefile ];then $(MAKE) -C ./source clean;fi;
+	if [ -f ./source/Makefile ];then $(MAKE) -C ./source clean distclean srcdir=./ builddir=./;fi;
 
diff -Naur samba/source/conf/smb.conf samba-brcm/source/conf/smb.conf
--- samba/source/conf/smb.conf	2010-09-14 19:35:20.000000000 -0300
+++ samba-brcm/source/conf/smb.conf	2018-11-24 14:14:21.827850768 -0400
@@ -16,15 +16,11 @@
 # This will prevent nmbd to search for NetBIOS names through DNS.
   dns proxy = no
 
-# Local Master setings
-#  domain master = no
-#  local master = Yes
-#  preferred master = Yes
-#  os level = 35
 
 # log files are not necessary,and can be ignored for now 
   log level = 0
-  log file = /var/samba/logsamba.%m
+# log file = /var/samba/logsamba.%m
+  log file = /dev/null
 
 # The specific set of interfaces / networks to bind to
   interfaces = br0
@@ -39,7 +35,10 @@
   map to guest = bad user
 
 # Performance related settings
-  socket options = TCP_NODELAY SO_RCVBUF=8192 SO_SNDBUF=8192
+  socket options = TCP_NODELAY SO_RCVBUF=65536 SO_SNDBUF=65536
+#  max xmit = 131072
+#  read size = 131072
+  use sendfile = yes
 #inactivity time out for client sessions 
   deadtime = 5
 
diff -Naur samba/source/conf/smb_telus.conf samba-brcm/source/conf/smb_telus.conf
--- samba/source/conf/smb_telus.conf	1969-12-31 20:00:00.000000000 -0400
+++ samba-brcm/source/conf/smb_telus.conf	2018-11-24 14:14:21.827850768 -0400
@@ -0,0 +1,71 @@
+#
+# Samba configuration script for Broadcom Linux Gateway
+#
+
+#======================= Global Settings =======================
+
+[global]
+
+# workgroup/NT-domain name your Samba server is part of
+  workgroup = WORKGROUP
+
+# server string is the equivalent of the NT Description field
+  netbios name = brcm-samba_test
+  server string = DSL Gateway
+
+# This will prevent nmbd to search for NetBIOS names through DNS.
+  dns proxy = no
+
+
+# log files are not necessary,and can be ignored for now
+  log level = 0
+# log file = /var/samba/logsamba.%m
+  log file = /dev/null
+
+# The specific set of interfaces / networks to bind to
+  interfaces = br0
+
+  load printers = no
+
+# security settings
+  security = user
+  guest account = nobody
+  #encrypt passwords = yes
+  #obey pam restrictions = yes
+  #fix bug QA-Bug #108531
+  wide links = no
+
+  #map to guest = bad user
+
+# Performance related settings
+  socket options = TCP_NODELAY SO_RCVBUF=65536 SO_SNDBUF=65536
+  use sendfile = yes
+#inactivity time out for client sessions
+  deadtime = 5
+
+
+#======================= Share Definitions =======================
+# USB Key folder
+[storage]
+  comment = All Storage devices
+  valid users = nobody
+  public = yes
+  writable = yes
+  browseable = no
+  path = /mnt/
+
+[homes]
+  comment = Home Directory
+# browseable=no here hides homes,but displays a scetion with username;(refer man smb.conf)
+  browseable = no
+  read only = no
+  create mask = 0777
+  directory mask = 0777
+  path = /var/samba/homes/%S
+  valid users = %S
+
+#[public]
+#  comment = shared folders on each volume
+#  browseable = yes
+#  read only = no
+#  path = /mnt
diff -Naur samba/source/lib/util.c samba-brcm/source/lib/util.c
--- samba/source/lib/util.c	2010-09-14 02:49:26.000000000 -0300
+++ samba-brcm/source/lib/util.c	2018-11-24 14:14:21.895819471 -0400
@@ -1198,6 +1198,14 @@
 	static fstring dnshostname;
 	struct hostent *hp;
 
+#if 1
+   /*this was causing session setup failure when there is no
+   * reachable dns server, for now there is no use of this name resolution.
+   * Just disable it for now
+   */       
+	return False;
+#endif
+
 	if (!*dnshostname) {
 		/* get my host name */
 		if (gethostname(dnshostname, sizeof(dnshostname)) == -1) {
diff -Naur samba/source/smbd/vfs.c samba-brcm/source/smbd/vfs.c
--- samba/source/smbd/vfs.c	2010-09-14 02:49:26.000000000 -0300
+++ samba-brcm/source/smbd/vfs.c	2018-11-24 14:14:22.015764245 -0400
@@ -514,7 +514,9 @@
 	release_level_2_oplocks_on_change(fsp);
 	DEBUG(10,("vfs_set_filelen: ftruncate %s to len %.0f\n", fsp->fsp_name, (double)len));
 	flush_write_cache(fsp, SIZECHANGE_FLUSH);
-	if ((ret = SMB_VFS_FTRUNCATE(fsp, fsp->fh->fd, len)) != -1) {
+
+/* modified by brcm */
+	/*if ((ret = SMB_VFS_FTRUNCATE(fsp, fsp->fh->fd, len)) != -1) {
 		set_filelen_write_cache(fsp, len);
 		notify_fname(fsp->conn, NOTIFY_ACTION_MODIFIED,
 			     FILE_NOTIFY_CHANGE_SIZE
@@ -522,7 +524,8 @@
 			     fsp->fsp_name);
 	}
 
-	return ret;
+	return ret;*/
+	return 0;
 }
 
 /****************************************************************************
